### PR TITLE
PAAS-2122 add dependencies check script

### PR DESCRIPTION
**Pull request by laurentfufu**
JIRA issue: https://jira.jahia.org/browse/PAAS-2122

Short description:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,6 +34,125 @@ Here are some rules that should be observed:
...
+							[api] is a jelastic keyword with args
+		for manifest './packages/one-shot/paas-1982-fix-jessionid.yml, section 'events':
+		no events detected
+		for manifest './packages/one-shot/paas-1982-fix-jessionid.yml, section 'onInstall':
+			[checkJahiaVersion] is an action from ['./packages/one-shot/paas-1982-fix-jessionid.yml']
+			[temporaryWorkaround] is an action from ['./packages/one-shot/paas-1982-fix-jessionid.yml']
+			[foreach] is a keyword
+				[temporaryWorkaround] is an action from ['./packages/one-shot/paas-1982-fix-jessionid.yml']
+        ```
+		If you called action can't be found, you will have something like this:
```
Comment (kevinvandel) [inline README.md:65]: Typo in "If you called action can't be found", but I'm not sure were... "your" ?

File: deps.py [hunk 2]
```diff
@@ -0,0 +1,977 @@
...
+from rich.table import Table
+from rich.progress import track
+
+
+console = Console()
+
+
+
+
+parser = argparse.ArgumentParser()
```
Comment (kevinvandel) [inline deps.py:20]: PEP 8: E303 too many blank lines

File: deps.py [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+subparser_graph.add_argument("file", nargs='*', type=argparse.FileType('r'), help="manifest to graph", default=None)
+
+
+subparser_check.add_argument("file", nargs='*', type=argparse.FileType('r'), help="manifest to check", default=None)
+subparser_check.add_argument("-q", "--quiet", help="be quiet", default=False, action='store_true')
+
+
+
+
+subparser_search.add_argument('-i', '--id', help="id number")
```
Comment (kevinvandel) [inline deps.py:43]: PEP 8: E303 too many blank lines

File: deps.py [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+subparser_search.add_argument('--not_legit', help="action is not legit", default=False, action="store_true")
+
+
+args = parser.parse_args()
+
+
+
+
+
+graph_manifest_parser = cmd2.Cmd2ArgumentParser()
```
Comment (kevinvandel) [inline deps.py:60]: PEP 8: E303 too many blank lines

File: deps.py [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+
+
+check_manifest_parser = cmd2.Cmd2ArgumentParser()
+check_manifest_parser.add_argument('file', type=str, help='the manifest to check')
+
+
+
+
+
+folder_list = []
```
Comment (kevinvandel) [inline deps.py:74]: PEP 8: E303 too many blank lines

File: deps.py [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+        "manifest": {
+            "concentrate": "true",
+            "bgcolor": "lemonchiffon",
+            "constraint": "false",
+            },
+        "section_onInstall": {
+            "concentrate": "true",
+            "bgcolor": "lavender",
+            },
+            "constraint": "false",
```
Comment (kevinvandel) [inline deps.py:157]: I guess `constraint` should be under `section_onInstall`

File: deps.py [hunk 7]
```diff
@@ -0,0 +1,977 @@
...
+            "style": "filled,dashed",
+            "color": "red",
+            "fillcolor": "lightpink",
+            "constraint": "false",
+            },
+        }
+
+graph = pgv.AGraph(**graph_attributes["root"])
+
+class Action():
```
Comment (kevinvandel) [inline deps.py:183]: PEP 8: E302 expected 2 blank lines, found 1

File: deps.py [hunk 8]
```diff
@@ -0,0 +1,977 @@
...
+    def content(self):
+        return self.__content
+    def parameters(self):
+        return self.__params
+    def from_file(self) -> str:
+        return self.__from_file
+
+
+
+class Manifest():
```
Comment (kevinvandel) [inline deps.py:204]: PEP 8: E303 too many blank lines (3)

File: deps.py [hunk 9]
```diff
@@ -0,0 +1,977 @@
...
+        self.__content = kwargs["content"]
+        self.__from_file = kwargs["from_file"]
+        try:
+            self.__params = kwargs["params"]
+        except:
+            pass
+
+    def name(self) -> str:
+        return self.__name
+    def content(self):
```
Comment (kevinvandel) [inline deps.py:195]: PEP 8: E301 expected 1 blank line, found 0

File: deps.py [hunk 10]
```diff
@@ -0,0 +1,977 @@
...
+        return self.__content
+    def parameters(self):
```
Comment (kevinvandel) [inline deps.py:197]: PEP 8: E301 expected 1 blank line, found 0

File: deps.py [hunk 11]
```diff
@@ -0,0 +1,977 @@
...
+        return self.__params
+    def from_file(self) -> str:
```
Comment (kevinvandel) [inline deps.py:199]: PEP 8: E301 expected 1 blank line, found 0

File: deps.py [hunk 12]
```diff
@@ -0,0 +1,977 @@
...
+    def kind(self) -> list:
+        return self.__kind
+
+    def full_content(self):
+        return self.__full_content
+
+    def mixins(self) -> list:
+        return self.__mixins
+
+    def onInstall(self) -> list:
```
Comment (kevinvandel) [inline deps.py:249]: Not snake_case :/

File: deps.py [hunk 13]
```diff
@@ -0,0 +1,977 @@
...
+    def add_called_by(self, called_by: list):
+        self.__called_by = list(set([*self.__called_by, *called_by]))  # also remove duplicates
+    def set_kind(self, kind: str):
+        self.__kind = kind
+    def set_section(self, section: str):
+        self.__section = section
+    def set_legit(self, legit: bool):
+        self.__legit = legit
+
+def update_node(node_id: int, **kwargs):
```
Comment (kevinvandel) [inline deps.py:312]: PEP 8: E302 expected 2 blank lines, found 1

File: deps.py [hunk 14]
```diff
@@ -0,0 +1,977 @@
...
+    def __init__(self, name, **kwargs):
+        self.__name = name
+        self.__parent = kwargs.get('parent', None)
+        self.__childs = kwargs.get('childs', [])
+        self.__call = kwargs.get('call', [])
+        self.__called_by = kwargs.get('called_by', [])
+        self.__kind = kwargs.get('kind', None)
+        self.__section = kwargs.get('section', None)
+        self.__legit = kwargs.get('legit', True)
+    def get_all_attributes(self) -> dict:
```
Comment (kevinvandel) [inline deps.py:269]: PEP 8: E301 expected 1 blank line, found 0
(same for all other methods below)

File: deps.py [hunk 15]
```diff
@@ -0,0 +1,977 @@
...
+    if kind:
+        nodes_list[node_id].set_kind(kind)
+    if section:
+        nodes_list[node_id].set_section(section)
+    if legit:
+        nodes_list[node_id].set_legit(legit)
+
+
+
+def get_node_parents_id_by_name(name: str) -> list:
```
Comment (kevinvandel) [inline deps.py:338]: PEP 8: E303 too many blank lines (3)

File: deps.py [hunk 16]
```diff
@@ -0,0 +1,977 @@
...
+    return [ nodes_list.index(x) for x in nodes_list if x.name() == name ]
+
+def search_node(get="object", **criteria) -> list:
```
Comment (kevinvandel) [inline deps.py:341]: PEP 8: E302 expected 2 blank lines, found 1

File: deps.py [hunk 17]
```diff
@@ -0,0 +1,977 @@
...
+        pass
+    elif not_legit:
+        prefix_with = ":x:"
+    elif not string:
+        console.rule("")
+        return
+    console.print(f"{prefix_with}{string}", **options)
+
+
+def isKeyword(word: str) -> bool:
```
Comment (kevinvandel) [inline deps.py:394]: Not snake_case :/

File: deps.py [hunk 18]
```diff
@@ -0,0 +1,977 @@
...
+        return [nodes_list.index(x) for x in nodes_list if x.name() == name]
+    except:
+        return ""
+
+
+def crawl(item, degree=1, section="", manifest_name="", previous_was_legit=True,
+        previous_degree=1, previous_item_id=0, called_by=""):
+    padding = ' '
+    width = degree * 4
+    manifest=manifest_list[get_manifest_id_by_name(manifest_name)]
```
Comment (kevinvandel) [inline deps.py:449]: PEP 8: E225 missing whitespace around operator

File: deps.py [hunk 19]
```diff
@@ -0,0 +1,977 @@
...
+            mixin = re.sub(r"^/\W*", "./", mixin)
+        mixin_node = search_node(name=mixin)[0]
+        all_action_nodes += mixin_node.childs()
+
+    if isinstance(item, dict):
+        legit=False
+        to_break=False
+        i=0
+        for k in item:
+            i+=1
```
Comment (kevinvandel) [inline deps.py:466]: PEP 8: E225 missing whitespace around operator

File: deps.py [hunk 20]
```diff
@@ -0,0 +1,977 @@
...
+
+
+def crawl(item, degree=1, section="", manifest_name="", previous_was_legit=True,
+        previous_degree=1, previous_item_id=0, called_by=""):
+    padding = ' '
+    width = degree * 4
+    manifest=manifest_list[get_manifest_id_by_name(manifest_name)]
+
+    # get all related actions nodes
+    all_action_nodes = []
```
Comment (kevinvandel) [inline deps.py:452]: It seems this variable is never used

File: deps.py [hunk 21]
```diff
@@ -0,0 +1,977 @@
...
+                parent_id = search_node(get="id", name=manifest_name, kind="manifest")
+                nodes_list.append(Node(k, parent=parent_id, kind=kind, legit=legit, section=section))
+            elif first_word in [x.name() for x in actions_list]:
+                matching_actions = [x for x in actions_list if x.name() == first_word]
+                # select the actions used in case of actions with the same name
+                for x in matching_actions:
+                    if x.from_file() == manifest_name:
+                        matching_actions = [x]
+                    elif x.from_file() in manifest.mixins():
+                        matching_actions = [x]
```
Comment (kevinvandel) [inline deps.py:497]: :thinking:

File: deps.py [hunk 22]
```diff
@@ -0,0 +1,977 @@
...
+                            parent=parent_id, section=section)[0]
+                    update_node(parent_id, add_childs=[node_id])
+                update_node(node_id, legit=legit, section=section)
+
+                if called_by:
+                    node_id = search_node(get="id", name=first_word, kind=kind,
+                            parent=parent_id, section=section)[0]
+                    if node_id != called_by:
+                        update_node(node_id, add_called_by=[called_by])
+                        update_node(called_by, add_call=[node_id])
```
Comment (kevinvandel) [inline deps.py:521]: First arg of `update_node` is supposed to be an int, but `called_by` is a string

File: deps.py [hunk 23]
```diff
@@ -0,0 +1,977 @@
...
+                            previous_degree=degree, previous_item_id=i, called_by=node_id)
+
+            elif degree > 1 and previous_item_id != 0:
+                legit=True
+                kind="parameter"
+                log(rf"{padding :>{width}}\[{k}] is a parameter")
+            elif not legit and previous_was_legit:
+                legit=False
+                to_break=True
+                kind="action"
```
Comment (kevinvandel) [inline deps.py:553]: PEP 8: E225 missing whitespace around operator

File: deps.py [hunk 24]
```diff
@@ -0,0 +1,977 @@
...
+                else:
+                    next_item = item[k]
+
+                if not isinstance(item[k], dict):
+                    crawl(next_item, degree = degree + 1, section=section, manifest_name=manifest_name,
+                            previous_degree=degree, previous_item_id=i, called_by=node_id)
+
+            elif degree > 1 and previous_item_id != 0:
+                legit=True
+                kind="parameter"
```
Comment (kevinvandel) [inline deps.py:548]: `kind` is never used here

File: deps.py [hunk 25]
```diff
@@ -0,0 +1,977 @@
...
+                    if node_id != called_by:
+                        update_node(node_id, add_called_by=[called_by])
+                        update_node(called_by, add_call=[node_id])
+
+            else:
+                legit=False
+                log(rf"{padding :>{width}}\[{k}] is an unknown kind, this should never be showned")
+
+        if not isinstance(item[k], str):
+            if not to_break:
```
Comment (kevinvandel) [inline deps.py:573]: Should be done in a single `if`

File: deps.py [hunk 26]
```diff
@@ -0,0 +1,977 @@
...
+
+                if section == "actions":
+                    nodes_list.append(Node(k, parent=parent_id, kind=kind, legit=legit, section=section))
+
+                if called_by:
+                    node_id = search_node(get="id", name=first_word, kind=kind,
+                            parent=parent_id, section=section)[0]
+                    if node_id != called_by:
+                        update_node(node_id, add_called_by=[called_by])
+                        update_node(called_by, add_call=[node_id])
```
Comment (kevinvandel) [inline deps.py:566]: First arg of `update_node` is supposed to be an int, but `called_by` is a string

File: deps.py [hunk 27]
```diff
@@ -0,0 +1,977 @@
...
+            except:
+                nodes_list.append(Node(first_word, parent=parent_id, kind=kind, legit=legit, section=section))
+                node_id = search_node(get="id", name=first_word, kind=kind,
+                        parent=parent_id, section=section)[0]
+                update_node(parent_id, add_childs=[node_id])
+            update_node(node_id, legit=legit, section=section)
+
+            if called_by and node_id != called_by:
+                update_node(node_id, add_called_by=[called_by])
+                update_node(called_by, add_call=[node_id])
```
Comment (kevinvandel) [inline deps.py:618]: First arg of `update_node` is supposed to be an int, but `called_by` is a string

File: deps.py [hunk 28]
```diff
@@ -0,0 +1,977 @@
...
+            if not search_node(name=first_word, kind=kind, parent=parent_id, section=section):
+                nodes_list.append(Node(first_word, parent=parent_id, kind=kind, legit=legit, section=section))
+
+            if called_by:
+                node_id = search_node(get="id", name=first_word, kind=kind,
+                        parent=parent_id, section=section)[0]
+
+                if node_id != called_by:
+                    update_node(node_id, add_called_by=[called_by])
+                    update_node(called_by, add_call=[node_id])
```
Comment (kevinvandel) [inline deps.py:652]: First arg of `update_node` is supposed to be an int, but `called_by` is a string

File: deps.py [hunk 29]
```diff
@@ -0,0 +1,977 @@
...
+    except:
+        return False
+
+    manifest_subgraph = graph.add_subgraph(name=f"cluster_manifest_{manifest.name()}", label=f"{manifest.name()}", **graph_attributes["manifest"])
+
+    for mixin in manifest.mixins():
+        graph_by_manifest(mixin)
+
+    # get all related actions nodes
+    all_action_nodes = []
```
Comment (kevinvandel) [inline deps.py:725]: `all_action_nodes` is never used

File: deps.py [hunk 30]
```diff
@@ -0,0 +1,977 @@
...
+                    target_node_id_name = nodes_list[target_node_id].name()
+                    target_parent_manifest_id = nodes_list[target_node_id].parent()
+                    target_parent_manifest_name = nodes_list[target_parent_manifest_id].name()
+                    target_node_name = f"{target_parent_manifest_name}_actions_{target_node_id_name}"
+                    graph.add_edge(node_name, target_node_name)
+
+
+def graph_manifest(name: str, output=None, to_terminal=False):
+    if re.search(regex_dict["is_mixins"], name):
+        other_mixins = [ x.name() for x in manifest_list if re.search(regex_dict["is_mixins"], x.name()) and x.name() != name ]
```
Comment (kevinvandel) [inline deps.py:764]:
```suggestion
        other_mixins = [x.name() for x in manifest_list if re.search(regex_dict["is_mixins"], x.name()) and x.name() != name]
```

File: deps.py [hunk 31]
```diff
@@ -0,0 +1,977 @@
...
+        for mixin in other_mixins:
+            crawl_by_manifest(mixin)
+            graph_by_manifest(mixin)
+
+    crawl_by_manifest(name)
+    graph_by_manifest(name)
+
+    mixins_list =  [ node for node in graph.nodes() if re.search(regex_dict["is_mixins"], node.name) and not re.search(rf"{name}", node.name)]
```
Comment (kevinvandel) [inline deps.py:772]:
```suggestion
    mixins_list = [node for node in graph.nodes() if re.search(regex_dict["is_mixins"], node.name) and not re.search(rf"{name}", node.name)]
```

File: deps.py [hunk 32]
```diff
@@ -0,0 +1,977 @@
...
+    for mixin in mixins_list:
+        actions[mixin] = set([ x.name() for x in actions_list if x.from_file() == mixin ])
+
+    all_actions = [ action for mixin in actions.keys() for action in actions[mixin] ]
+
+    duplicated = set([action for action in all_actions if all_actions.count(action) > 1 ])
+
+    for action in duplicated:
+        from_file = [ file for file in actions.keys() if action in actions[file] ]
+        log(f"'{action}' is duplicated: {from_file}", not_legit=True)
```
Comment (kevinvandel) [inline deps.py:810]: PEP 8: E201 whitespace after '['
PEP 8: E202 whitespace before ']'

File: deps.py [hunk 33]
```diff
@@ -0,0 +1,977 @@
...
+        self.quiet = False
+        self.add_settable(cmd2.Settable('quiet', bool, 'be quiet', self))
+
+        self.console = False
+        self.add_settable(cmd2.Settable('console', bool, 'output generated dot to console', self))
+
+    @cmd2.with_argparser(graph_manifest_parser)
+    def do_graph(self, opts):
+        """Generate dot format of actions dependencies tree for a given manifest"""
+        SILENT_MODE = opts.quiet
```
Comment (kevinvandel) [inline deps.py:836]: `SILENT_MODE` is never used

File: deps.py [hunk 34]
```diff
@@ -0,0 +1,977 @@
...
+            elif isinstance(value, int):
+                value = str(value)
+            str_item.append(value)
+        rows.append(str_item)
+
+    for row in rows:
+        table.add_row(*row)
+
+
+    console.print(table)
```
Comment (kevinvandel) [inline deps.py:898]: PEP 8: E303 too many blank lines (2)

File: deps.py [hunk 35]
```diff
@@ -0,0 +1,977 @@
...
+
+
+def check_if_not_legit():
+    not_legit = search_node(get="id", legit=False)
+    if not_legit:
+        display_not_legit(not_legit)
+        sys.exit(1)
+
+if __name__ == "__main__":
```
Comment (kevinvandel) [inline deps.py:907]: PEP 8: E305 expected 2 blank lines after class or function definition, found 1

**Review by kevinvandel — CHANGES_REQUESTED** [review]

File: deps.py (hunk 22 above)
Comment (laurentfufu) [reply deps.py:521]: `called_by` is now `None` by default in the crawl `crawl` function

**Review by kevinvandel — APPROVED** [review]

**Pull request merged**